### PR TITLE
feat: Implement Device System Events

### LINKED ISSUE #4099: Core Metadata publish Device System Events
# 🚀 Feature Request

### Relevant Package [**REQUIRED**]

 This feature request is for Core Metadata

### Description [**REQUIRED**]
Implement publishing of Device System Events per System Events ADR (Link TBD)

### Describe the solution you'd like
See ADR

### Describe alternatives you've considered
See ADR

### PR DESCRIPTION
**Pull request by lenny-intel**
Also, refactored to use common Messaging Bootstrap Handler

closes #4099

Implements [related ADR](https://docs.edgexfoundry.org/2.3/design/adr/0024-system-events) 

Dependent on the following PRs:
- https://github.com/edgexfoundry/go-mod-core-contracts/pull/752
- https://github.com/edgexfoundry/go-mod-bootstrap/pull/360

Signed-off-by: Leonard Goodell <leonard.goodell@intel.com>

**If your build fails**  due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/edgex-go/blob/main/.github/Contributing.md

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] I am not introducing a breaking change (if you are, flag in conventional commit message with `BREAKING CHANGE:` describing the break)
- [x] I am not introducing a new dependency (add notes below if you are)
- [x] I have added unit tests for the new feature or bug fix (if not, why?)
- [x] I have fully tested (add details below) this the new feature or bug fix (if not, why?)
- [x] I have opened a PR for the related docs change (if not, why?)
  link TBD

## Testing Instructions
Run edgex stack with Device Virtual
Stop Core Metadata container
Build and run this branch locally (require branches from PR referenced above) with DEBUG enabled
Using curl or postman delete, add and remove Device Virtual devices (can't use EdgeX UI from docker since core-metadata is running locally)
Verify log show debug message that System Event was published

### CODE REVIEW

File: internal/core/metadata/application/device.go
Comment (cloudxxx8) [inline internal/core/metadata/application/device.go:271]: use `errors.CommonEdgeX` instead of normal error

File: internal/core/metadata/application/device.go
Comment (cloudxxx8) [inline internal/core/metadata/application/device.go:291]: the correlation id should retrieve from Context, and it will be the same one as the changing device request for debugging

File: internal/core/metadata/application/device.go
Comment (cloudxxx8) [inline internal/core/metadata/application/device.go:69]: why not using async call?

**Review by cloudxxx8 — CHANGES_REQUESTED** [review]

File: internal/core/metadata/application/device.go
Comment (cloudxxx8) [reply internal/core/metadata/application/device.go]: we don't handle the error and we usually use async call to publish data to message bus.
if the connection to the message bus is blocking due to network issue or something, the Device Add/Update/Delete will be blocked, too.
that's why we use `go addDeviceCallback(ctx, dic, device)` to prevent any unexpected situation.

Comment (lenny-intel) [reply internal/core/metadata/application/device.go]: OK

Comment (lenny-intel) [reply internal/core/metadata/application/device.go]: Added go func's

File: internal/core/metadata/application/device.go
Comment (cloudxxx8) [inline internal/core/metadata/application/device.go:195]: since it's asnyc, we might be able to remove this comment

File: internal/core/metadata/application/device.go
Comment (cloudxxx8) [inline internal/core/metadata/application/device.go:196]: log could add more context here:
`lc.Errorf("issue happens when publishing the device update system event: %v", err)`

File: internal/core/metadata/application/device.go
Comment (cloudxxx8) [inline internal/core/metadata/application/device.go:98]: log could add more context here:
`lc.Errorf("issue happens when publishing the device add system event: %v", err)`

File: internal/core/metadata/application/device.go
Comment (cloudxxx8) [inline internal/core/metadata/application/device.go:281]: you might need to create the error in this way:
declare the error with a normal error in line 273:
`var noMessagingClientError = goErrors.New("unable to publish Device System Event: MessageBus Client not available")`

and wrap the error for return here
`return errors.NewCommonEdgeXWrapper(noMessagingClientError)`

In this way, the errors.CommonEdgeX can record which line in which file cause this error in the debug message.
Users can still use `errors.As` to determine whether the error contains `noMessagingClientError`

**Review by cloudxxx8 — CHANGES_REQUESTED** [review]

File: internal/core/metadata/application/device.go
Comment (lenny-intel) [reply internal/core/metadata/application/device.go:196]: That is already added in the function. I have it returning error in order to validate in unit tests. Think it is better to just have the function log the error.

File: internal/core/metadata/application/device.go
Comment (lenny-intel) [reply internal/core/metadata/application/device.go:281]: Ok, sound good.

File: internal/core/metadata/application/device.go
Comment (lenny-intel) [reply internal/core/metadata/application/device.go:196]: I have reworked the error handling so this is no longer needed.

File: internal/core/metadata/application/device.go
Comment (lenny-intel) [reply internal/core/metadata/application/device.go]: No longer needed.

File: internal/core/metadata/application/device.go
Comment (lenny-intel) [reply internal/core/metadata/application/device.go]: No loner returning error, just logging it.

File: internal/core/metadata/application/device.go
Comment (lenny-intel) [reply internal/core/metadata/application/device.go]: No longer needed

**Review by cloudxxx8 — DISMISSED** [review]
LGTM

**Review by farshidtz — CHANGES_REQUESTED** [review]
The snap fails to start because security-secretstore-setup exits with non-zero code. The CI logs currently don't get the logs when the snap fails to even start (we are working on improving this).

I downloaded the build from Github (see for example [here](https://github.com/edgexfoundry/edgex-go/actions/runs/2750254562); scroll down) and installed manually to check the logs:
```
$ snap install --dangerous ./edgexfoundry_2.3.0-dev.30_amd64.snap
Run service command "start" for services ["consul" "core-command" "core-data" "core-metadata" "kong-daemon" "postgres" "redis" "security-bootstrapper-redis" "security-consul-bootstrapper" "security-proxy-setup" "security-secretstore-setup" "vault"] of snap "edgexfoundry" (systemctl command [start snap.edgexfoundry.security-secretstore-setup.service] failed with exit status 1: Job for snap.edgexfoundry.security-secretstore-setup.service failed because the control process exited with error code.

$ sudo journalctl -n 1000 | grep edgexfoundry | grep ERROR
level=ERROR ts=2022-07-28T08:14:23.57437856Z app=security-secretstore-setup source=init.go:329 msg="token provider failed: /snap/edgexfoundry/x1/bin/security-file-token-provider terminated with non-zero exit code 1"
```
Unfortunately, security-secretstore-setup is not printing the error message. Unrelated to this PR, it looks like standard output and standard error aren't captured by EdgeX logger [here](https://github.com/edgexfoundry/edgex-go/blob/062af8d3cd30019d44081b1021c6648b4b920d9b/internal/security/secretstore/tokenprovider.go#L86-L95). [This code](https://github.com/edgexfoundry/edgex-go/blob/062af8d3cd30019d44081b1021c6648b4b920d9b/internal/security/secretstore/execrunner.go#L36-L41) sets to use os output streams but not sure if this is ever called.

I could not figure out why the changes in this PR are causing this failure.

Full logs: [edgexfoundry_2.3.0-dev.30.logs.txt](https://github.com/edgexfoundry/edgex-go/files/9207796/edgexfoundry_2.3.0-dev.30.logs.txt)

**Comment by lenny-intel** [comment]
@farshidtz , @jim-wang-intel create this PR to test which go-mod and which version has triggered this. We have narrowed it down to go-mod-bootstrap v2.2.1-dev.9
https://github.com/edgexfoundry/edgex-go/pull/4105

I am taking a look at the commit for that to see how it could be impacting this.

**Comment by lenny-intel** [comment]
@farshidtz , that change to go-mod-bootstrap was to make sure the exit code it set properly when bootstrapping fails. Before it always returned with 0 exit code. 

I suspect that the issue with Secret Store setup in the test was always there, be now has the proper exit code.

**Comment by lenny-intel** [comment]
@farshidtz , found the culprit !

The file token provider is incorrectly returning `false` in its bootstrap hander which now causes it to exit with non-zero exit code.

We'll get a fix in for that which will un-block this PR.

**Comment by lenny-intel** [comment]
@jim-wang-intel , @farshidtz , Rebased this PR to pull in fixes from the other PR.

**Comment by lenny-intel** [comment]
Put in draft to it doesn't get merged until this PR is merged otherwise TAF tests will fail.
https://github.com/edgexfoundry/edgex-compose/pull/261#pullrequestreview-1055010670

**Review by jim-wang-intel — APPROVED** [review]

**Pull request merged**
